### PR TITLE
Update for typescript-eslint v6

### PR DESCRIPTION
**Pull request by jtbandes**
### Public-Facing Changes

- Update package version to 1.0.0 since the package has been stable for quite a while.

**Breaking changes:**

- Updated peer and dev dependencies to v6 of typescript-eslint packages
- Updated `plugin:@foxglove/typescript` config to be based on new `strict-type-checked` and `stylistic-type-checked` configs from upstream
  - This adds several new rules that may require disabling or fixing.
- Deleted `no-private-identifier` rule (see #46)
- Enabled `prefer-hash-private` by default

### Description

See typescript-eslint v6 release notes: https://typescript-eslint.io/blog/announcing-typescript-eslint-v6/

- `checkCompoundAssignments: true` is now the default for `restrict-plus-operands`
- Switched from `@typescript-eslint/recommended` + `@typescript-eslint/recommended-requiring-type-checking` to  `@typescript-eslint/strict-type-checked` + `@typescript-eslint/stylistic-type-checked`
  - `prefer-nullish-coalescing` and `prefer-optional-chain` are now on by default
- Switched to `@typescript-eslint/rule-tester` for `RuleTester`

### CODE REVIEW

**Comment by jtbandes** [comment]
Currently looking for feedback on these changes and whether the following newly enabled rules (which cause a lot of errors in Studio) should be left up to the downstream user to decide whether they want them on or off?

```
"@typescript-eslint/consistent-type-definitions": off
"@typescript-eslint/consistent-indexed-object-style": off
"@typescript-eslint/consistent-generic-constructors": off
"@typescript-eslint/no-confusing-void-expression": off
"@typescript-eslint/dot-notation": off
"@typescript-eslint/prefer-for-of": off
"@typescript-eslint/no-dynamic-delete": off
"@typescript-eslint/array-type": off
"@typescript-eslint/no-invalid-void-type": off
"@typescript-eslint/no-unsafe-enum-comparison": off
"@typescript-eslint/unified-signatures": off
@typescript-eslint/no-unnecessary-type-arguments
@typescript-eslint/prefer-function-type
@typescript-eslint/no-redundant-type-constituents
@typescript-eslint/no-base-to-string
@typescript-eslint/prefer-nullish-coalescing
@typescript-eslint/prefer-optional-chain (think we already had this but it catches more cases now)
@typescript-eslint/prefer-reduce-type-parameter
@typescript-eslint/no-throw-literal
@typescript-eslint/prefer-string-starts-ends-with
@typescript-eslint/non-nullable-type-assertion-style
```

**Comment by bryfox** [comment]
> Currently looking for feedback on these changes

I'd be OK leaving these disabled: `no-base-to-string` and `prefer-nullish-coalescing` seemed to cause false positives; `consistent-type-definitions`, `array-type`, and perhaps `prefer-optional-chain` can (IMO) harm readability and/or IDE usability.

I would most favor re-enabling `no-unsafe-enum-comparison` and `no-throw-literal`.

For the rest, from the point of view of Console, I think they'd be straightforward to fix in a single PR if we wanted them enabled, but I wouldn't override the defaults to do so; I'd rather keep as close to Studio as possible.

**Comment by foxymiles** [comment]
[consistent-type-definitions](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/consistent-type-definitions.md)
[consistent-indexed-object-style](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/consistent-indexed-object-style.md)
[consistent-generic-constructors](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/consistent-generic-constructors.md)
[no-confusing-void-expression](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-confusing-void-expression.md)
[dot-notation](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/dot-notation.md)
[prefer-for-of](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-for-of.md)
[no-dynamic-delete](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-dynamic-delete.md)
[array-type](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/array-type.md)
[no-invalid-void-type](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-invalid-void-type.md)
[no-unsafe-enum-comparison](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-unsafe-enum-comparison.md)
[unified-signatures](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/unified-signatures.md)
[no-unnecessary-type-arguments](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-unnecessary-type-arguments.md)
[prefer-function-type](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-function-type.md)
[no-redundant-type-constituents](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-redundant-type-constituents.md)
[no-base-to-string](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-base-to-string.md)
[prefer-nullish-coalescing](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-nullish-coalescing.md)
[prefer-optional-chain](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-optional-chain.md)
[prefer-reduce-type-parameter](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-reduce-type-parameter.md)
[no-throw-literal](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-throw-literal.md)
[prefer-string-starts-ends-with](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-string-starts-ends-with.md)
[non-nullable-type-assertion-style](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/non-nullable-type-assertion-style.md)

Rules for reference

**Comment by foxymiles** [comment]
[no-confusing-void-expression](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-confusing-void-expression.md)

[no-invalid-void-type](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-invalid-void-type.md)

These two seem possibly useful. The rest not so much.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@foxglove/eslint-plugin",
-  "version": "0.23.0",
+  "version": "1.0.0",
```
Comment (amacneil) [inline package.json:3]: what a time to be alive

**Review by bryfox — APPROVED** [review]
The override choices seem sensible to me, and from the point of view of console (https://github.com/foxglove/console/pull/3900), I'm 👍

**Pull request merged**